### DISCUSSION TITLE
Where to transform output for Eldoc?

### DISCUSSION THREAD

**Discussion by angrybacon** [joaotavora/eglot#1278] [Q&A]
Some documentation coming from the server response can contain newline characters:

```lisp
(:kind "markdown" :value "\n```typescript\nconst test: Test\n```\n")
```

Due to how that is prettified with `gfm-view-mode`, that results in an unwanted blank line in the echo area

<img width="778" alt="Screenshot 2023-08-28 at 03 02 55" src="https://github.com/joaotavora/eglot/assets/5608624/c40b7082-a88e-42c6-8e96-d0bd264a8eaa">

Although I haven't quite found out yet where the opening line of the code block is being handled seeing as there's no empty line _above_ the documentation in the echo area, only _below_ it

Here is how I currently nuke that empty line but I'm guessing this could breaks other parts of the involved libraries so it stays an undocumented hack in my own configuration for now

```lisp
(advice-add 'eglot--format-markup :filter-return
  (lambda (output)
    (replace-regexp-in-string (rx "\n```" eos) "" output))
  '((name . me/eglot--trim-block-end)))
```

You can reproduce the issue with the following code, mostly taken from `eglot--format-markup`:

````lisp
(message
 (with-temp-buffer
   (let ((markdown-fontify-code-blocks-natively t)
         (inhibit-message t)
         (message-log-max nil))
     (insert "
```typescript
const test: Test
```
")
     (ignore-errors (delay-mode-hooks (gfm-view-mode)))
     (font-lock-ensure)
     (string-trim (buffer-string)))))
````

Can you see a more elegant way to deal with this?

**Comment by joaotavora** [comment]
Mathieu Marques ***@***.***> writes:

> (advice-add 'eglot--format-markup :filter-return
>   (lambda (output)
>     (replace-regexp-in-string (rx "\n```" eos) "" output))
>   '((name . me/eglot--trim-block-end)))

Well you can try to set 'eglot-prefer-plaintext' to t, and hopefully the
server will give you plaintext, not markdown.

But other than that, unfortunately, no, not at the moment.  Well, short
of writing a completely new member to eldoc-display-functions, which
would be 100% legal and supported, but also too verbose.  That makes
those 4 lines look at least decent.

Maybe I should just bite the bullet and export 'eglot--format-markup' to
'eglot-format-markup' and enshrine tweaks such as yours.

For context, some time ago, in the Emacs bug tracker, some related discussion.

bug#61373, bug#61072 and bug#62029

Not sure I'm misremembering, but what's needed (and more elegant) is
some kind of configurable markdown renderer that the user can optimize
for space/compatcness.  But it doesn't exist yet.

João

**Reply by angrybacon** [reply]
> Well you can try to set 'eglot-prefer-plaintext' to t, and hopefully the
server will give you plaintext, not markdown.

Good to know thanks, I don't have this in my version yet. But I do like the Markdown version, just not how the end of a block behaves in the echo area

Thanks for the references I'll have a look
